### PR TITLE
Specify both this and relative positions when setting computeAt

### LINKED ISSUE #477: Domain mapping in setThisComputeAtAxis
https://github.com/csarofeen/pytorch/blob/023f4775a6129d5b1108d0a5a4348716c068d955/torch/csrc/jit/codegen/cuda/tensor_view.cpp#L240-L248

Not sure if this is considered a bug or not. It may not cause any actual problems, but here's an apparent inconsistency in computeAt position settings.

Example fusion:
```
  auto tv0 = makeSymbolicTensor(1);
  fusion.addInput(tv0);
  auto tv1 = broadcast(tv0, {true, true, false});
  auto tv2 = broadcast(tv1, {true, false, false, false});
  auto tv3 = makeSymbolicTensor(4);
  fusion.addInput(tv3);
  auto tv4 = add(tv2, tv3);
  fusion.addOutput(tv4);

  tv0->computeAt(tv4, 2);
```

The math looks fine:
```
%kernel_math {
T1[ bS1{1}, bS2{1}, iS0{i1} ] compute_at( T2, 2 ) = broadcast( T0[ iS0{i1} ] )
T2[ bS3{1}, bS1{1}, bS2{1}, iS0{i1} ] compute_at( T4, 2 ) = broadcast( T1[ bS1{1}, bS2{1}, iS0{i1} ] compute_at( T2, 2 ) )
T4[ iS9{i9}, iS10{i11}, iS11{i13}, iS8{i1} ]
   = T2[ bS3{1}, bS1{1}, bS2{1}, iS0{i1} ] compute_at( T4, 2 )
   + T3[ iS4{i9}, iS5{i11}, iS6{i13}, iS7{i15} ];
}
```

However,  the `this` position of `tv1` is set as 2 because of the way `setThisComputeAtAxis` maps domains. Semantically speaking, it should be 1. This is related to the root mapping with broadcast axes, where we can correctly identify actual mappings by using the broadcast flags. However, in this case, the domains are not root, so the flags can't be used directly.

The codegen seems to work fine, though. Here's the generated code:

```
__global__ void CUDAGeneratedKernel(Tensor<float, 1> T0, Tensor<float, 4> T3, Tensor<float, 4> T4) {
  for(size_t ki26 = 0; ki26 < T3.size[0]; ++ki26) {
    for(size_t ki27 = 0; ki27 < T3.size[1]; ++ki27) {
      float T1[T0.size[0]];
      float T2[T0.size[0]];
      for(size_t ki28 = 0; ki28 < T0.size[0]; ++ki28) {
        T1[ki28]
           = T0[(ki28 * T0.stride[0])];
      }
      for(size_t ki39 = 0; ki39 < T0.size[0]; ++ki39) {
        T2[ki39]
           = T1[ki39];
      }
      for(size_t ki46 = 0; ki46 < T3.size[2]; ++ki46) {
        for(size_t ki47 = 0; ki47 < T0.size[0]; ++ki47) {
          T4[(ki26 * T4.stride[0]) + (ki27 * T4.stride[1]) + (ki46 * T4.stride[2]) + ki47]
            = T2[ki47]
            + T3[(ki26 * T3.stride[0]) + (ki27 * T3.stride[1]) + (ki46 * T3.stride[2]) + (ki47 * T3.stride[3])];
        }
      }
    }
  }
}
```

### PR DESCRIPTION
**Pull request by naoyam**
The problem is that TensorView::setThisComputeAtAxis may return a wrong
position when a new broadcast axis is added to a TensorDomain that already
has broadcast axes. It may not correctly disambiguate which broadcast
axes of the producer should match with one in the consumer. Root mapping
had a similar problem, which was resolved by using the broadcast flag
list, but that list is only applicable to root domains.

Instead of fixing setThisComputeAtAxis, this commit changes all the
uses of setComputeAt to pass both the this and relative positions so
that setThisComputeAtAxis is no longer needed. As far as I see, each
use case of setComputeAt does know the value of this position (some
use cases needed minor changes), so all of them are replaced with the
another interface of setComputeAt.

Fixes #477

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 1]
```diff
@@ -582,17 +555,19 @@ TensorView* TensorView::cache_after() {
...
         if (output->hasComputeAt()) {
-          TransformReplay::replayPasC(consumer, output, -1);
           auto output_ca_pos = output->getThisComputeAtAxis();
-          consumer->setComputeAt(output, output_ca_pos);
+          auto this_pos =
+              TransformReplay::replayPasC(consumer, output, output_ca_pos)
+                  .second;
+          consumer->setComputeAt(output, this_pos, output_ca_pos);
         }
       }
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp:597]: This part may need more work. See #483. In this PR, I tried to minimize changes in the current behavior.

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp [hunk 2]
```diff
@@ -144,6 +144,23 @@ Expr* Val::getOrigin() const {
   return fusion_->origin(this);
 }
 
+bool Val::isProducer(const Val* other) const {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp]: Nit, maybe we should name this as:
isConsumerOf(other)
and
isProducerOf(other)

isConsumer/Producer isn't an inherent property of Val, but a relation between two vals, the syntax just seems a little more clear to me, but don't have a strong opinion of you object

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.h [hunk 3]
```diff
@@ -209,6 +209,12 @@ class TORCH_CUDA_API Val : public Statement {
   // was found
   Expr* getOrigin() const;
 
+  //! Returns true when a given Val is a producer of this Val
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: nit on comments for clarity:
`Returns true when other is a producer of this`

File: torch/csrc/jit/codegen/cuda/ir_interface_nodes.h [hunk 4]
```diff
@@ -322,8 +322,6 @@ class TORCH_CUDA_API TensorView : public Val {
     domain_ = td;
   }
 
-  void setComputeAt(TensorView* computeAtView, int axis);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_interface_nodes.h:325]: 👍

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 5]
```diff
@@ -152,36 +152,40 @@ IterDomain* TensorView::axis(int pos) const {
...
+      thisPos > 0 && (unsigned)thisPos <= nDims(),
+      "Invalid this computeAt position for T",
+      name(),
+      ": ",
+      thisPos);
+  // When computeAtView is a consumer, the CA axes must not include
+  // reductions. Note that an output tensor may be set as computed at
+  // another output tensor even if they are not a producer and a
+  // consumer.
+  if (isConsumer(computeAtView)) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp]: Do you want to validate else both are outputs?

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 1 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp:597]: Makes sense.

**Review by csarofeen — APPROVED** [review]
Looks good, stamping. Just some minor comments.

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp (hunk 2 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp]: Thanks. Makes sense. Updated.

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 5 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp]: That's a good idea. Validation added.

**Pull request merged**